### PR TITLE
add scikit-umfpack

### PR DESCRIPTION
**Pull request by grlee77**
This is a recipe for scikit-umfpack.  I have skipped CI because it depends on #682 which has not been merged.

### CODE REVIEW

**Comment by jakirkham** [comment]
cc @rgommers

**Comment by rgommers** [comment]
Nice timing @grlee77! Coincidence, or was this in response to the discussion on the scipy optional dependencies?

CircleCI is failing with:

    conda.resolve.NoPackagesFound:  Package missing in current linux-64 channels: 
      - suitesparse

**Comment by rgommers** [comment]
Ah, I see gh-655 for adding suitesparse.

**Comment by jakirkham** [comment]
Do you know if scikit-umfpack needs dynamic libraries for SuiteSparse or will it work with static libraries, @rgommers?

**Comment by rgommers** [comment]
I've only used it with dynamic libs, and this text 

    Unfortunately, SuiteSparse’s makefiles do not support building a shared library out
    of the box. You may find Stefan Fürtinger instructions helpful.

on https://pypi.python.org/pypi/scikit-umfpack seems to suggest static is a problem, but I'm not sure. 

Cc'ing @rc as ``scikit-umfpack`` author.

**Comment by rc** [comment]
The link to Stefan Fürtinger instructions on pypi is broken. See https://github.com/scikit-umfpack/scikit-umfpack for the correct link. As for the issue, the instructions have been added by @timtadh, cc'ing him, as I have no experience with building it myself - I am using only the debian/*ubuntu suitesparse version.

**Comment by timtadh** [comment]
@rgommers I did not have any luck getting things to work with SuiteSparse as a static library. I found lots of other people having similar problems for: Matlab, Octave, etc.... Basically, SuiteSparse is the defacto standard for high performance sparse matrix work. However, the build system is pretty funky and not setup to build shared objects. So you have to modify it. I attempted to do that myself and made a few mistakes and then stumbled on Stefan's directions. Using them and my intuition I got it working. Here is the link: http://fuertinger.lima-city.de/research.html#building-numpy-and-scipy 

I have only done this once and it was about a year ago now. I luckily have not had to do it again. I recall it took me about 2 working days to get everything setup.

**Comment by grlee77** [comment]
re: @rgommers 
partly coincidence:  I had made this recipe (and #682) for local use about a month ago, but thought I would clean it up and upload here after seeing the spicy package discussion.  The default build gives shared libraries in the most recent Suitesparse release.  I plan to revisit the shared build as in #682 in the suitesparse feedstock once #655 is merged.

**Comment by grlee77** [comment]
This is now ready for review.  

`bld.bat` was add for eventual windows support, but we don't currently have a windows build for the `suitesparse` dependency.

@timtadh, @pv, @rc : Are there any plans to release a new version of `scikit-umfpack` in the near future so that the patches used here would not be needed?  Both address issues that have already been resolved in the master branch (see the comments in the patches section of `meta.yaml`).  Also, if any of you interested in being a maintainer on this recipe let me know and I will gladly add you.

File: recipes/scikit-umfpack/bld.bat [hunk 1]
```diff
@@ -0,0 +1,6 @@
+
+set UMFPACK=%PREFIX%\lib
```
Comment (jakirkham) [inline recipes/scikit-umfpack/bld.bat]: So, I know Windows support doesn't exist *yet*. Normally libraries are place in `%LIBRARY_LIB%`, which is just short form for `%PREFIX%\Library\lib`.

File: recipes/scikit-umfpack/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  name: scikit-umfpack
+  version: {{ version }}
+
+source:
+  fn: scikit-umfpack--{{ version }}.tar.gz
+  url: https://pypi.io/packages/source/s/scikit-umfpack/scikit-umfpack-{{ version }}.tar.gz
+  sha256: 9ebfb4ac4cd1ec545a9211160c65129a77b6e2fab02d8cfa30b10278f5773fd1
+  patches:
+    # osx patch below is needed in release 0.2.1, but was fixed by commit 987af60 in master
+    - setup-osx.patch  # [osx]
```
Comment (jakirkham) [inline recipes/scikit-umfpack/meta.yaml]: Could we just replace this with the patch generated from commit ( https://github.com/scikit-umfpack/scikit-umfpack/commit/987af60 )? That way there will be no special casing required.

Comment (jakirkham) [inline recipes/scikit-umfpack/meta.yaml]: Not sure if you know this or not, but you can tack on `.patch` or `.diff` to any commit or PR to get the corresponding patch or diff (e.g. https://github.com/scikit-umfpack/scikit-umfpack/commit/987af60.patch ).

**Comment by jakirkham** [comment]
Thanks for the thoughts on handling SuiteSparse @rgommers, @rc, and @timtadh. I think we are going to take the path of least resistance here and try embracing SuiteSparse 4.5.x. Let's see how that goes. That version includes dynamic libraries and it was recently added here to conda-forge.

File: recipes/scikit-umfpack/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    - py3.patch
+
+build:
+  number: 0
+  skip: True  # [win]
+
+requirements:
+  build:
+    - python
+    - numpy
```
Comment (jakirkham) [inline recipes/scikit-umfpack/meta.yaml]: As this links against `numpy`, we need to add `x.x` here.

File: recipes/scikit-umfpack/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    - scipy
+    - suitesparse
+    - swig
+    - setuptools
+  run:
+    - python
+    - numpy
```
Comment (jakirkham) [inline recipes/scikit-umfpack/meta.yaml]: Also, add `x.x` here too.

File: recipes/scikit-umfpack/meta.yaml (hunk 2 above)
Comment (grlee77) [inline recipes/scikit-umfpack/meta.yaml]: thanks. I didn't know you could get the patch like that.  I had been using `diff -u`.  Pretty sure I initially made the patch here and later found the PR that had already fixed it.

File: recipes/scikit-umfpack/py3.patch [hunk 5]
```diff
@@ -0,0 +1,33 @@
+--- scikits/umfpack/interface.py	2015-09-11 16:19:05.000000000 -0400
++++ cikits/umfpack/interface_new.py	2016-05-27 01:40:48.000000000 -0400
```
Comment (jakirkham) [inline recipes/scikit-umfpack/py3.patch]: `cikits` -> `scikits`?

**Comment by grlee77** [comment]
Forgot to skip appveyor last time, but otherwise this is passing.  Windows support will await SuiteSparse on windows which may be a bit off as we need OpenBLAS etc there first.

Also, the license for these wrappers are BSD 3-clause, but the underlying UMFPACK library from SuiteSparse is GPL v2 which is perhaps the more relevant constraint for other projects wanting to use this.  Should that be noted in the license section?

**Comment by jakirkham** [comment]
What version of UMFPACK are we using (in other words what version is SuiteSparse providing)?

**Comment by grlee77** [comment]
UMFPACK 5.7.6 (May 4, 2016) is the version included in SuiteSparse 4.5.3.

**Comment by rgommers** [comment]
> Also, the license for these wrappers are BSD 3-clause, but the underlying UMFPACK library from SuiteSparse is GPL v2 

Yeah, we removed direct linking to UMFPACK from Scipy due to license concerns - the BSD license seems questionable here.

> which is perhaps the more relevant constraint for other projects wanting to use this. Should that be noted in the license section?

Maybe as a comment only? I'd think that the license info here should reflect what the package license states, not the concerns we may have about that.

**Comment by jakirkham** [comment]
So, UMFPACK has been under various licenses over time. Depending on what versions of UMFPACK they support, it may be possible for scikit-umfpack to be under BSD 3-Clauses in some cases.

However, the copy we are using (and linking to) says in no uncertain terms it is GPL 2. Given that we are shipping a binary that is linked to this GPL 2 copy of UMFPACK, I think we have to say GPL 2 for this binary. We can note that scikit-umfpack has a different license for the source code, but we are in the business of distributing binaries and need to be very clear about how they are built and what can be done with them.

I ran into this exact situation with pyFFTW. As there is a commercial license for FFTW in addition to the open source GPL 2 version, it is possible to have pyFFTW be a binary that is BSD 3-Clause licensed legally. In this situation, I stated that it could be GPL 2 or BSD 3-Clause. This situation seems similar to me. Not sure if that is the most clear way to describe as it makes it seem like dual licensing, but it at least alerts the user to the GPL license so they can decide whether they want to investigate or not. If others have better ideas, I would be curious to hear.

@msarahan, do you know of any other relevant metadata fields that we could use to handle this sort of situation?

**Comment by grlee77** [comment]
I changed the license to GPL 2 as discussed above.

not sure why I had to add setuptools to the runtime requirements when it had previously passed CI without it

**Comment by jjhelmus** [comment]
> not sure why I had to add setuptools to the runtime requirements when it had previously passed CI without it

We recently added a configuration to the  .condarc file that conda-forge used to not automatically include pip and in turn setuptools when python is specified as a dependency.  Now these requirements need to be explicitly specified in the recipe.

**Comment by jakirkham** [comment]
LGTM. Thoughts anyone else?

**Comment by rgommers** [comment]
LGTM too.

**Comment by jjhelmus** [comment]
LGTM, merging.

**Comment by jakirkham** [comment]
Thanks everyone. The repo for this recipe can be found here ( https://github.com/conda-forge/scikit-umfpack-feedstock ).

**Pull request merged**